Authentication methods & database backend

WeirdAlex03 opened the pull request:
Adds full backend functionality for signup and login

This doesn't give visible feedback to the user. It should, but that's not part of the testing procedure requirements so we're not gonna worry about that today

Closes #17, closes #20

@sumanyuj please don't close and delete this one >:(

WeirdAlex03 committed:
Account models and signup/login code

WeirdAlex03 committed:
Adding routing for auth wip

WeirdAlex03 committed:
fix wrong name for pwd conf

sumanyuj commented:
I am running into a error with registration. When I click the "Post" button I get a 500 Error, and is has something to do with a false CSRF token apparently. Also this DB error: 

database-1  | 2024-10-25 02:41:14.339 UTC [34] ERROR:  relation "eg_app_account" does not exist at character 89
database-1  | 2024-10-25 02:41:14.339 UTC [34] STATEMENT:  SELECT "eg_app_account"."id", "eg_app_account"."email", "eg_app_account"."pwdHash" FROM "eg_app_account" WHERE "eg_app_account"."email" = 'sumanyuj@buffalo.edu' LIMIT 21
django-1    | [25/Oct/2024 02:41:14] "POST /register HTTP/1.1" 500 145

sumanyuj commented:
> I am running into a error with registration. When I click the "Post" button I get a 500 Error, and is has something to do with a false CSRF token apparently. Also this DB error:
> 
> database-1 | 2024-10-25 02:41:14.339 UTC [34] ERROR: relation "eg_app_account" does not exist at character 89 database-1 | 2024-10-25 02:41:14.339 UTC [34] STATEMENT: SELECT "eg_app_account"."id", "eg_app_account"."email", "eg_app_account"."pwdHash" FROM "eg_app_account" WHERE "eg_app_account"."email" = '[sumanyuj@buffalo.edu](mailto:sumanyuj@buffalo.edu)' LIMIT 21 django-1 | [25/Oct/2024 02:41:14] "POST /register HTTP/1.1" 500 145

and yeah, I did run makemigrations

WeirdAlex03 commented:
can you give exact reproduction steps? i didn't see that particular error while working on it

WeirdAlex03 commented:
and where does CSRF come into that? the db error doesnt mention that

sumanyuj commented:
Sure. 

1) run docker compose force create and all of that jazz
2) put in email and "Password@1234"
3) Confirm, press validate button. All Green. 
4) Press post button
5) Redirect to 500 Error page

sumanyuj commented:
> and where does CSRF come into that? the db error doesnt mention that

![Screenshot by Dropbox Capture](https://github.com/user-attachments/assets/b092108f-1787-4cdc-ae2f-0573a68bdb30)
![Screenshot by Dropbox Capture](https://github.com/user-attachments/assets/c3261253-16e6-47d4-a14e-900249a0b252)

CSRF cookies do not match.

WeirdAlex03 reviewed (commented):
## My steps:

### Attempt 1:
- dcupbfr
- register with your exact email and password
- validate, then post
- no error on frontend or backend, redirected as normal
- reviewing network logs after submitting, i did also see `csrftoken` and `csrfmiddlewaretoken` mismatch
- **only partially reproduced**

### Attempt 2:
- docker system prune
- docker volume prune
- dcupbfr
- unintentionally but conviently forget to run migrations
- clear cookies on browser and refresh
- note value of `csrftoken` in Cookies and `csrfmiddlewaretoken` input in form HTML before submitting. They are different.
- register with your exact email and password
- validate, then post
- Get "Server Error (500)" on front end and equivlent errors to yours in logs
- The value of `csrftoken` in Cookies is the same as it was before; `csrfmiddlewaretoken` in Request matches the same named input on the form
- **successfully reproduced**

### Attempt 3:
- docker system prune
- docker volume prune
- dcupbfr
- docker compose exec django python manage.py makemigrations
- docker compose exec django python manage.py migrate
- clear cookies on browser and refresh
- note value of `csrftoken` in Cookies and `csrfmiddlewaretoken` input in form HTML before submitting. They are different.
- register with your exact email and password
- validate, then post
- no error on frontend or backend, redirected as normal
- The value of `csrftoken` in Cookies is the same as it was before; `csrfmiddlewaretoken` in Request matches the same named input on the form
- **only partially reproduced**

### Attempt 4
- same setup as Attempt 3, but running *only* migrate *without* makemigrations
- same result as Attempt 2 (errors on front and backend)

## My conclusions
- The `csrftoken` cookie and `csrfmiddlewaretoken` form field appear to be unrelated to eachother
- Make sure you run **both** makemigrations **and** migrate. Running only the latter will apply some, but not all migrations needed for it to work
- Issue: Closed; user error (but we really do need to figure out how to get migrations to run automatically so this isn't a problem for the TAs. thats for a different PR though)

sumanyuj commented:
Did not run migrations, sorry about that. 

I am not getting the 500 error, however, when I login, I am unable to see the Auth-Token as a response to the login http header.

WeirdAlex03 commented:
Well you're not seeing the Auth-Token cookie because I set it as a header instead. I also forgot a few other things about that on the handout, I guess it's back to the IDE to me

WeirdAlex03 committed:
Refactor to use Django auth, add logout

WeirdAlex03 committed:
remove now-unnecessary models

WeirdAlex03 committed:
drop debug prints

sumanyuj commented:
> Well you're not seeing the Auth-Token cookie because I set it as a header instead. I also forgot a few other things about that on the handout, I guess it's back to the IDE to me

It's not in the header either, will check again tho.

Eric-Butcher commented on evergreen/eg_app/views.py:
```diff
@@ -33,6 +34,57 @@ def validate(request):
+            return HttpResponseRedirect(root)
+
+        # Now confirmed valid, create account
+        try:
```
Is there a way to do this with normal control flow? Using try/except control flow should generally only be used for _exceptional_ control flow. Here you are expecting an exception and using it for program logic. 

This is not as bad of a thing to do in Python since Python's implementation of this does not necessarily slow down the code, it is just a smell.

Eric-Butcher commented on evergreen/eg_app/views.py:
```diff
@@ -33,6 +34,57 @@ def validate(request):
+    return HttpResponseBadRequest()
+
+def login_view(request: HttpRequest):
+    root = '/'
```
You have repeated yourself here. Make this a constant variable in the global file scope. ROOT_PATH: str = "/"

Eric-Butcher commented on evergreen/eg_app/views.py:
```diff
@@ -33,6 +34,57 @@ def validate(request):
+        try:
+            User.objects.get(email=email)
+        except User.DoesNotExist:
+            newAcct = User.objects.create_user(email, email, password)
```
add username = email to be more clear about why you are passing in two email paramaters

WeirdAlex03 committed:
Address feedback

- Refactor register dupe check to avoid try/catch
- Add explicit kwargs to `create_user` call in `register`
- Make root path a global var for all `views` methods

sumanyuj reviewed (approved):
LGTM
